### PR TITLE
Add a faster custom parser for checkpoint metadata

### PR DESCRIPTION
**Pull request by apang-ns**
The existing configuration parser used by WiredTiger is slow. This is not a problem generally. For example, it is used to read configuration out of mongod.conf on startup, and being slow by microseconds on startup is not a problem.

Unfortunately, this config parser is also used for reading checkpoint metadata out of files. Reading checkpoint metadata is in the critical path of preparing to checkpoint (__checkpoint_prepare()), where we take schema and dhandle locks and are getting ready to perform a checkpoint. During this time, application threads handling RW requests are slowed down or stopped. This happens every 60 seconds.

For an environment with a large number of dhandles, the amount of checkpoint metadata config that needs to be parsed is large enough to be a performance problem.

In our local testing, parsing this metadata can take 15 seconds in a 30 minute interval. This means that parsing operations is preventing application threads from working 0.8% of the time.

The custom parsing code in this PR aims to reduce this performance overhead. In our testing, this parser is 6 times faster.

The code here is limited to improving performance in this critical path and does not solve the general slowness in mongo's config library.

### CODE REVIEW

File: src/third_party/wiredtiger/src/txn/txn_ckpt.c [hunk 1]
```diff
@@ -1335,8 +1335,17 @@ __checkpoint_lock_dirty_tree(WT_SESSION_IMPL *session,
 	WT_ASSERT(session, !need_tracking ||
 	    WT_IS_METADATA(dhandle) || WT_META_TRACKING(session));
 
-	/* Get the list of checkpoints for this file. */
-	WT_RET(__wt_meta_ckptlist_get(session, dhandle->name, &ckptbase));
+	/*
+	 * Get the list of checkpoints for this file.
+	 * Use the custom checkpoint metadata parser if configured.
+	 */
+	if (F_ISSET(conn, WT_CONN_CKPT_METADATA_FAST_PARSER)) {
```
Comment (kand-netskope) [inline src/third_party/wiredtiger/src/txn/txn_ckpt.c]: I think we could do this inside of __wt_meta_ckptlist_get() and avoid duplicating the beginning and end of that function.
If the issue is to get conn in that function, S2C should work.

File: src/third_party/wiredtiger/src/meta/meta_ckpt.c [hunk 2]
```diff
@@ -240,6 +243,246 @@ __ckpt_compare_order(const void *a, const void *b)
...
+	while ((found = strstr(found, "WiredTigerCheckpoint."))) {
+		WT_ERR(__wt_realloc_def(
+			session, &allocated, slot + 1, &ckptbase));
+		ckpt = &ckptbase[slot];
+
+		/*
+		 * Length of "WiredTigerCheckpoint." is 21, start searching
+		 * for the value there.
+		 */
+		name_end = strchr(found + 21, '=');
```
Comment (kand-netskope) [inline src/third_party/wiredtiger/src/meta/meta_ckpt.c]: Maybe use (sizeof("WiredTigerCheckpoint.") - 1) or create a defined value instead of 21.

File: src/third_party/wiredtiger/src/meta/meta_ckpt.c [hunk 3]
```diff
@@ -240,6 +243,246 @@ __ckpt_compare_order(const void *a, const void *b)
...
+		WT_RET(__wt_strndup(session, found,
+			(int)(name_end - found), &ckpt->name));
+
+		/*
+		 * Begin parsing attributes starting from the end of the
+		 * name string.
+		 */
+		__ckpt_parse_attributes(session, name_end, ckpt);
+
+		found++;
```
Comment (kand-netskope) [inline src/third_party/wiredtiger/src/meta/meta_ckpt.c]: I think we could start the next iteration at least at name_end.

File: src/third_party/wiredtiger/src/meta/meta_ckpt.c [hunk 4]
```diff
@@ -240,6 +243,246 @@ __ckpt_compare_order(const void *a, const void *b)
...
+		name_end = strchr(found + 21, '=');
+
+		WT_RET(__wt_strndup(session, found,
+			(int)(name_end - found), &ckpt->name));
+
+		/*
+		 * Begin parsing attributes starting from the end of the
+		 * name string.
+		 */
+		__ckpt_parse_attributes(session, name_end, ckpt);
```
Comment (kand-netskope) [inline src/third_party/wiredtiger/src/meta/meta_ckpt.c]: We could probably start the checkpoint parsing at name_end + 1 and it makes it clearer that we start after the '='.

File: src/third_party/wiredtiger/src/meta/meta_ckpt.c [hunk 5]
```diff
@@ -240,6 +243,246 @@ __ckpt_compare_order(const void *a, const void *b)
...
+{
+	const char *str_p = *s;
+	const char *prefix_p = prefix;
+
+	/*
+	 * While we have not reached the end of the prefix string or the search
+	 * string, and while we have not exited the current level of
+	 * configuration indicated by parens, look for a prefix match.
+	 */
+	while (*str_p != '\0' && *str_p != ')' && *prefix_p != '\0') {
```
Comment (kand-netskope) [inline src/third_party/wiredtiger/src/meta/meta_ckpt.c]: I wonder if this would be easier with strtok or strpbrk. Have you looked at those ?
It seems like you could use them to find = and then ",\")"

File: src/third_party/wiredtiger/src/meta/meta_ckpt.c [hunk 6]
```diff
@@ -240,6 +243,246 @@ __ckpt_compare_order(const void *a, const void *b)
...
 }
 
+/*
+ * __ckpt_str_find_not_alphanum --
+ *  Return the first character that is not alphanumeric.
+ */
+const char *
+__ckpt_str_find_not_alphanum(const char *s)
+{
+	while (*s != '\0' && __wt_isalnum(*s))
```
Comment (kand-netskope) [inline src/third_party/wiredtiger/src/meta/meta_ckpt.c]: Do we expect anything else than ',' or ')' ?

File: src/third_party/wiredtiger/src/txn/txn_ckpt.c (hunk 1 above)
Comment (apang-ns) [reply src/third_party/wiredtiger/src/txn/txn_ckpt.c]: Hm ok. I was trying to keep the original code the same, so if anything goes wrong and we flip the flag it's as good as a revert. I can refactor some of this and try to keep it low risk.

File: src/third_party/wiredtiger/src/meta/meta_ckpt.c (hunk 2 above)
Comment (apang-ns) [reply src/third_party/wiredtiger/src/meta/meta_ckpt.c]: Ok

File: src/third_party/wiredtiger/src/meta/meta_ckpt.c (hunk 5 above)
Comment (apang-ns) [reply src/third_party/wiredtiger/src/meta/meta_ckpt.c]: Hm yea that might be less code. I'll try it and see if it's similar in speed.

File: src/third_party/wiredtiger/src/meta/meta_ckpt.c (hunk 3 above)
Comment (apang-ns) [reply src/third_party/wiredtiger/src/meta/meta_ckpt.c]: Yes

File: src/third_party/wiredtiger/src/meta/meta_ckpt.c (hunk 4 above)
Comment (apang-ns) [reply src/third_party/wiredtiger/src/meta/meta_ckpt.c]: Sure

File: src/third_party/wiredtiger/src/txn/txn_ckpt.c (hunk 1 above)
Comment (kand-netskope) [reply src/third_party/wiredtiger/src/txn/txn_ckpt.c]: I see. The drawback is if we patch other versions, we have to check if anything else changed.

I think you could do something light in __wt_meta_ckplist_get() like:
if (F_ISSET(S2C(session), WT_CONN_CKPT_METADATA_FAST_PARSER))
    __wt_meta_ckptlist_get_custom()
else {
    current code
}

File: src/third_party/wiredtiger/src/meta/meta_ckpt.c (hunk 5 above)
Comment (apang-ns) [reply src/third_party/wiredtiger/src/meta/meta_ckpt.c]: I rewrote it with the library functions. Performance is more or less the same (slower by 60ms after 30 minutes).

**Comment by apang-ns** [comment]
Unit tests pass both in the original and custom paths.

Checkpoint metadata parsing correctness tested by hand.

Need to look into the integration tests to see if edge cases for checkpoint metadata are covered.

File: src/third_party/wiredtiger/src/meta/meta_ckpt.c [hunk 7]
```diff
@@ -240,6 +244,217 @@ __ckpt_compare_order(const void *a, const void *b)
...
+
+	WT_RET(__wt_strndup(session, name, (int)(*s - name), &ckpt->name));
+	/*
+	 * Find the checkpoint metadata config string inside
+ 	 * the parentheses.
+	 *
+	 * Advance past "=" to look for the closing parenthesis.
+	 */
+	(*s)++;
+	paren = strpbrk(*s + 1, ")(");
```
Comment (kand-netskope) [inline src/third_party/wiredtiger/src/meta/meta_ckpt.c:438]: Did you mean to search for both ")(" or just ")" ?

File: src/third_party/wiredtiger/src/meta/meta_ckpt.c [hunk 8]
```diff
@@ -240,6 +244,217 @@ __ckpt_compare_order(const void *a, const void *b)
...
+
+	/*
+	 * Sanity check that the parentheses exist and that there
+	 * are no nested parentheses.
+	 */
+	if (**s != '(' || *paren != ')')
```
Comment (kand-netskope) [inline src/third_party/wiredtiger/src/meta/meta_ckpt.c]: *paren might be a problem if strpbrk couldn't find anything since paren would be NULL

File: src/third_party/wiredtiger/src/meta/meta_ckpt.c (hunk 7 above)
Comment (kand-netskope) [inline src/third_party/wiredtiger/src/meta/meta_ckpt.c:438]: Do we need to be safe and check for *s being NULL before moving further ?

File: src/third_party/wiredtiger/src/meta/meta_ckpt.c [hunk 9]
```diff
@@ -240,6 +244,217 @@ __ckpt_compare_order(const void *a, const void *b)
...
+{
+	uint64_t len;
+
+	/* It is possible for the addr to be empty. */
+	len = 0;
+	if (val != NULL)
+		len = strlen(val);
+
+	switch (attr_idx) {
+		case 0:
```
Comment (kand-netskope) [inline src/third_party/wiredtiger/src/meta/meta_ckpt.c]: It might be clearer to have some defines for each value that represent the case like:
#define CKPT_ATTR_ADDR 0

After that, you could also initialize the attribute array with this to make it even clearer:
http://gcc.gnu.org/onlinedocs/gcc-4.1.2/gcc/Designated-Inits.html

File: src/third_party/wiredtiger/src/txn/txn_ckpt.c (hunk 1 above)
Comment (apang-ns) [reply src/third_party/wiredtiger/src/txn/txn_ckpt.c]: Ok I did what you suggested. Like you said it might make patching a bit trickier because the diff is not that clean.

One thing I was thinking about is to do:

```
if (F_ISSET()) {
    __wt_meta_ckptlist_get_custom();
    goto skip;
}
current code
skip:
   ...
```

This would make the diff cleaner, but is a bit unusual.

File: src/third_party/wiredtiger/src/meta/meta_ckpt.c (hunk 7 above)
Comment (apang-ns) [reply src/third_party/wiredtiger/src/meta/meta_ckpt.c:438]: I meant to have both ")(" here as a sanity check that there are no unexpected nested parenthesized sub configurations inside of the attributes list.

File: src/third_party/wiredtiger/src/meta/meta_ckpt.c (hunk 8 above)
Comment (apang-ns) [reply src/third_party/wiredtiger/src/meta/meta_ckpt.c]: Is there a case where we shouldn't be able to find the closing parenthesis? I thought that the checkpoint metadata always contained a "subconfig" with the list of metadata attributes.

Comment (apang-ns) [reply src/third_party/wiredtiger/src/meta/meta_ckpt.c]: Yes, I am expecting that we will always find the closing parenthesis unless there is a problem.

Is there a case where we would not find the matching closing parenthesis after the list of checkpoint metadata attributes? I believe that the checkpoint metadata looks like this:

WiredTigerCheckpoint.<int>=(attrkey1=attrval1, attrkey2=attrval2,...)

The existing code also expects the above format (__ckpt_load()), but it looks like the underlying config code won't bail if it finds the key before hitting the closing parenthesis.

Should we be more lenient here?

File: src/third_party/wiredtiger/src/meta/meta_ckpt.c (hunk 7 above)
Comment (apang-ns) [reply src/third_party/wiredtiger/src/meta/meta_ckpt.c:438]: Do you mean after L428, or before it?

There is a check in L418 for *s being NULL. After L418 I don't think it's possible for *s to be NULL right?

File: src/third_party/wiredtiger/src/meta/meta_ckpt.c (hunk 9 above)
Comment (apang-ns) [reply src/third_party/wiredtiger/src/meta/meta_ckpt.c]: Yes that is better. Let me make that change.

File: src/third_party/wiredtiger/src/meta/meta_ckpt.c (hunk 7 above)
Comment (kand-netskope) [reply src/third_party/wiredtiger/src/meta/meta_ckpt.c:438]: I think because you are moving *s further with (*s)++, it is not the same as in L418 anymore but I guess like for paren, we could assume that the string is correct if the rest of the code does also.

Comment (kand-netskope) [reply src/third_party/wiredtiger/src/meta/meta_ckpt.c:438]: Oh I see. That's why you have a check below for *paren != ')'.

File: src/third_party/wiredtiger/src/meta/meta_ckpt.c (hunk 8 above)
Comment (kand-netskope) [reply src/third_party/wiredtiger/src/meta/meta_ckpt.c]: OK. I guess if it was NULL, we would probably have even bigger problems.

**Review by kand-netskope — APPROVED** [review]

File: src/third_party/wiredtiger/src/meta/meta_ckpt.c (hunk 9 above)
Comment (apang-ns) [reply src/third_party/wiredtiger/src/meta/meta_ckpt.c]: This is done.

File: src/third_party/wiredtiger/src/meta/meta_ckpt.c (hunk 7 above)
Comment (apang-ns) [reply src/third_party/wiredtiger/src/meta/meta_ckpt.c:438]: That's right.

Comment (apang-ns) [reply src/third_party/wiredtiger/src/meta/meta_ckpt.c:438]: As long as *s + 1 is not NULL, it won't segfault and should find the closing parenthesis if it is there. I don't think *s can be NULL here since we are just incrementing *s by 2.

File: src/third_party/wiredtiger/src/meta/meta_ckpt.c (hunk 8 above)
Comment (apang-ns) [reply src/third_party/wiredtiger/src/meta/meta_ckpt.c]: Actually, you are right. I should check for parens being NULL before deferencing. If parent is NULL (and the checkpoint metadata string is corrupted), we want it to assert, not segfault.
